### PR TITLE
[ml/train] Training Interfaces [3/4]: `DataParallelTrainer` interface

### PR DESCRIPTION
**Pull request by amogkam**
Interface for `DataParallelTrainer` and updates to ScalingConfig definition.

Depends on #22986 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        train_dataset: Optional[GenDataset] = None,
+        extra_datasets: Optional[Dict[str, GenDataset]] = None,
+        preprocessor: Optional[Preprocessor] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+        raise NotImplementedError
+
+
+@dataclass
+class _DataParallelScalingConfig:
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py:63]: We ended up deciding to make this an internal only concept. Users are not expected to interact with this for now.

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    """Configuration for scaling data parallel training.
+
+    num_workers (int): The number of workers (Ray actors) to launch.
+        Each worker will reserve 1 CPU by default. The number of CPUs
+        reserved by each worker can be overridden with the
+        ``resources_per_worker`` argument.
+    use_gpu (bool): If True, training will be done on GPUs (1 per
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py:70]: It was originally was, but after some discussion, the push was to make this dataclass an internal only concept and not expose it as part of the API.

The API for scaling config is still just going to be a dict. The schema for the dict is going to be dependent on the Trainer it's being used for.

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+from ray.ml.checkpoint import Checkpoint
+from ray.train import BackendConfig
+from ray.tune import PlacementGroupFactory
+from ray.util.annotations import DeveloperAPI
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class DataParallelFunctionTrainer(FunctionTrainer):
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/ml/train/impl/data_parallel_trainer.py:18]: +1

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    """
+
+    num_workers: int
+    use_gpu: bool = False
+    resources_per_worker: Optional[Dict] = None
+    max_retries: int = 3
+    placement_strategy: str = "PACK"
+
+    def __post_init__(self):
+        raise NotImplementedError
```
Comment (richardliaw) [inline python/ray/ml/train/impl/data_parallel_trainer.py:91]: Do we need to have this here? is this part of dataclasses by default?

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            workers enable distributed communication, for example torch or
+            horovod. If no Backend should be set up, then set this to None.
+        scaling_config: Configuration for how to scale data parallel training.
+        run_config: Configuration for the execution of the training run.
+        train_dataset: Either a distributed Ray :ref:`Dataset <dataset-api>`
+            or a Callable that returns a Dataset, to use for training. If a
+            ``preprocessor`` is also provided, it will be fit on this
+            dataset and this dataset will be transformed.
+        extra_datasets: Any extra Datasets (such as validation or test
+            datasets) to use for training. If a ``preprocessor`` is
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py:40]: Right, this is just the interface PR. Does not include any implementation. This is just to make sure that the method signatures and abstractions look good.

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py:70]: Edit: there is now going to be a single schema for all trainers. We plan to revisit this for beta.

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py]: Changed!

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    This Trainer runs the provided function on multiple Ray Actors. The
+    provided datasets are automatically sharded to each Ray actor running
+    the training function.
+
+    Args:
+        train_func: The training function to execute.
+            This can either take in no arguments or a ``config`` dict.
+        train_func_config: Configurations to pass into
+            ``train_func`` if it accepts an argument.
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py:27]: It literally is just 1 argument though 🙂. See the docstring for `train_func` above, it should only take in 0 or 1 arguments.

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        run_config: Configuration for the execution of the training run.
+        train_dataset: Either a distributed Ray :ref:`Dataset <dataset-api>`
+            or a Callable that returns a Dataset, to use for training. If a
+            ``preprocessor`` is also provided, it will be fit on this
+            dataset and this dataset will be transformed.
+        extra_datasets: Any extra Datasets (such as validation or test
+            datasets) to use for training. If a ``preprocessor`` is
+            provided, the datasets specified here will only be transformed,
+            and not fit on.
+        preprocessor: A preprocessor to preprocess the provided datasets.
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py]: Updated the docstring to hopefully make this more clear. But no documentation exists for Preprocessor. Would there still be something to reference?

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 4 above)
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py]: removed

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        worker). Defaults to False. The number of GPUs reserved by each
+        worker can be overridden with the ``resources_per_worker``
+        argument.
+    resources_per_worker (Optional[Dict]): If specified, the resources
+        defined in this Dict will be reserved for each worker. The
+        ``CPU`` and ``GPU`` keys (case-sensitive) can be defined to
+        override the number of CPU/GPUs used by each worker.
+    max_retries (int): Number of retries when Ray actors fail.
+        Defaults to 3. Set to -1 for unlimited retries.
+    placement_strategy (str): The placement strategy to use for the
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py]: Good point, added!

File: python/ray/ml/config.py [hunk 9]
```diff
@@ -1,8 +1,58 @@
...
 # kwargs.
 # TODO(xwjiang): After Tuner is implemented, define the schema
 RunConfig = Dict[str, Any]
 
 ScalingConfig = Dict[str, Any]
+
+
+@dataclass
+class _ScalingConfigDataClass:
+    """Configuration for scaling training.
```
Comment (richardliaw) [inline python/ray/ml/config.py:17]: Maybe you can leave a comment saying that this is the schema for ScalingConfig, and in the future (after beta) this will be the default representation for ScalingConfigs

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    provided datasets are automatically sharded to each Ray actor running
+    the training function.
+
+    Args:
+        train_func: The training function to execute.
+            This can either take in no arguments or a ``config`` dict.
+        train_func_config: Configurations to pass into
+            ``train_func`` if it accepts an argument.
+        backend_config: Used to specify which backend to setup on the
+            workers enable distributed communication, for example torch or
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/ml/train/impl/data_parallel_trainer.py:29]: nit: consider "collective_communication_config" or something?

**Review by richardliaw — APPROVED** [review]
Left some nits but otherwise looks good.

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (maxpumperla) [reply python/ray/ml/train/impl/data_parallel_trainer.py:38]: @amogkam  right, I know. It's just hard to tell from these interfaces and their docs where we intend things to be implemented, is all

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (maxpumperla) [reply python/ray/ml/train/impl/data_parallel_trainer.py:29]: personally, I think the "backend" concept is a good one

**Review by ericl — COMMENTED** [review]
I expect a DataParallelTrainer to define a function taking the dataset sub-shard, and rank. Shouldn't this be in the interface?

Basically, for Trainer: you define the overall `train_loop`. For DataParallelTrainer, you define `train_loop_for_rank(shard, rank)`

**Comment by amogkam** [comment]
@ericl do you mind raising this up with the group?

I'm ok with that change, but would like to get consensus from other folks before making the change.

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py:29]: Keeping this as backend.

File: python/ray/ml/config.py [hunk 11]
```diff
@@ -1,8 +1,58 @@
...
 # Right now, RunConfig is just an arbitrary dict that specifies tune.run
 # kwargs.
 # TODO(xwjiang): After Tuner is implemented, define the schema
 RunConfig = Dict[str, Any]
 
 ScalingConfig = Dict[str, Any]
+
+
+@dataclass
+class _ScalingConfigDataClass:
```
Comment (ericl) [inline python/ray/ml/config.py:16]:
```suggestion
class ScalingConfig:
```

File: python/ray/ml/config.py [hunk 12]
```diff
@@ -1,8 +1,58 @@
...
+    """Configuration for scaling training.
+
+    num_workers (int): The number of workers (Ray actors) to launch.
```
Comment (ericl) [inline python/ray/ml/config.py:19]:
```suggestion
    num_workers: The number of workers (Ray actors) to launch.
```

Please remove types from docstrings per our style guide.

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+            provided, the datasets specified here will only be transformed,
+            and not fit on.
+        preprocessor: A ray.ml.preprocessor.Preprocessor to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A checkpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        train_func: Callable,
```
Comment (ericl) [inline python/ray/ml/train/impl/data_parallel_trainer.py:48]: Specify types of the callable?

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class DataParallelTrainer(Trainer):
+    """A Trainer for data parallel training.
+
+    This Trainer runs the provided function on multiple Ray Actors. The
+    provided datasets are automatically sharded to each Ray actor running
+    the training function.
```
Comment (ericl) [inline python/ray/ml/train/impl/data_parallel_trainer.py:21]: Can you add an example of what kind of train_func to provide? It's not clear what the sharding means. Does each func know its rank?

File: python/ray/ml/train/impl/data_parallel_trainer.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+from ray.ml.checkpoint import Checkpoint
+from ray.train import BackendConfig
+from ray.util.annotations import DeveloperAPI
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class DataParallelTrainer(Trainer):
+    """A Trainer for data parallel training.
```
Comment (ericl) [inline python/ray/ml/train/impl/data_parallel_trainer.py:17]: When do I use this class? Who should use it? Examples?

**Comment by amogkam** [comment]
This one isn't ready for re-review yet. Comments will be addressed after getting consensus on train_func API.

File: python/ray/ml/config.py (hunk 9 above)
Comment (amogkam) [reply python/ray/ml/config.py:17]: Added!

File: python/ray/ml/config.py (hunk 11 above)
Comment (amogkam) [reply python/ray/ml/config.py]: Never mind, had to undo this. ScalingConfig is already being used to define the type. When we make the dataclass the default representation, we can change the name accordingly.

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 13 above)
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py]: Added

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 14 above)
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py]: Updated

File: python/ray/ml/train/impl/data_parallel_trainer.py (hunk 15 above)
Comment (amogkam) [reply python/ray/ml/train/impl/data_parallel_trainer.py]: Updated

File: python/ray/ml/train/data_parallel_trainer.py [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+    ``train_loop_config`` will be passed in as the argument. This is useful if you
+    want to tune the values in ``train_loop_config`` as hyperparameters.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    ``ray.ml.train.TRAIN_DATASET_KEY``), then it will be split into multiple dataset
+    shards that can be consumed by ``train_loop_per_worker``. All the other datasets
+    will not be sharded.
+
+    Inside the ``train_loop_per_worker`` function, you can use any of the
+    :ref:`Ray Train function utils <train-api-func-utils>`
```
Comment (richardliaw) [inline python/ray/ml/train/data_parallel_trainer.py:50]:
```suggestion
    :ref:`Ray Train function utils <train-api-func-utils>`.
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+
+        class MyBackend(Backend):
+            def on_start(self, worker_group, backend_config):
+                def set_env_var(env_var_value):
+                    import os
+                    os.environ["MY_ENV_VAR"] = env_var_value
+
+                worker_group.execute(set_env_var, backend_config.env_var)
+
+        @dataclasses
```
Comment (amogkam) [inline python/ray/ml/train/data_parallel_trainer.py:142]:
```suggestion
        @dataclasse
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+        class MyDataParallelTrainer(DataParallelTrainer):
+            def __init__(self, *args, **kwargs):
+                predefined_train_loop_per_worker = lambda: 1
+                super().__init__(predefined_train_loop_per_worker, *args, **kwargs)
+
+
+    For 2, you can implement the ``ray.train.Backend`` and ``ray.train.BackendConfig``
+    interfaces.
+
+    ..code-block:: python
```
Comment (richardliaw) [inline python/ray/ml/train/data_parallel_trainer.py:129]:
```suggestion
    .. code-block:: python
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+        Example:
+            .. code-block:: python
+
+                import ray
+                from ray import train
+                from ray.ml.train import TRAIN_DATASET_KEY
+
+                def train_loop_for_worker():
+                    dataset_shard_for_this_worker = train.get_dataset_shard(
+                        TRAIN_DATASET_KEY)
```
Comment (richardliaw) [inline python/ray/ml/train/data_parallel_trainer.py:87]: can we perhaps get the TRAIN_DATASET by default?

Like, train_shard = train.get_dataset_shard()

File: python/ray/ml/train/data_parallel_trainer.py (hunk 18 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py]: fixed!

File: python/ray/ml/train/data_parallel_trainer.py (hunk 19 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py:87]: Hmm, I think it's better to be explicit here?

The user is already specifying `TRAIN_DATASET_KEY` in the init args. It might be better to have consistency when retrieving the dataset.

File: python/ray/ml/train/data_parallel_trainer.py [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+            ...
+
+    .. code-block:: python
+
+        def train_loop_per_worker(config: Dict):
+            ...
+
+    If ``train_loop_per_worker`` accepts an argument, then
+    ``train_loop_config`` will be passed in as the argument. This is useful if you
+    want to tune the values in ``train_loop_config`` as hyperparameters.
```
Comment (ericl) [inline python/ray/ml/train/data_parallel_trainer.py:42]: Maybe simplify to always take config?

File: python/ray/ml/train/data_parallel_trainer.py (hunk 19 above)
Comment (ericl) [reply python/ray/ml/train/data_parallel_trainer.py:87]: Would it be more readable to just use the string "train" instead of an enum constant? No strong opinion but I've seen this in other Python APIs.

File: python/ray/ml/trainer.py [hunk 21]
```diff
@@ -121,10 +123,11 @@ def training_loop(self):
...
             dataset and this dataset will be transformed.
-        extra_datasets: Any extra Datasets (such as validation or test
-            datasets) to use for training. If a ``preprocessor`` is
-            provided, the datasets specified here will only be transformed,
-            and not fit on.
+        datasets: Any Ray Datasets to use for training. Use
+            ``ray.ml.train.TRAIN_DATASET_KEY``. to denote which dataset is the training
+            dataset. If a ``preprocessor`` is provided and has not already been fit,
+            it will be fit on the training dataset. All datasets will be transformed
+            by the ``preprocessor`` if one is provided.
```
Comment (ericl) [inline python/ray/ml/trainer.py:130]: Did we decide only the train dataset will be transformed? (fine if the decision is to try everything, just want to make sure this wasn't a typo)

**Review by ericl — APPROVED** [review]
Two minor comments.

File: python/ray/ml/trainer.py (hunk 21 above)
Comment (amogkam) [reply python/ray/ml/trainer.py:130]: Hmm, I thought only train dataset will be fit, but everything will be transformed (validation dataset for example would need to be transformed still).

File: python/ray/ml/train/data_parallel_trainer.py (hunk 19 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py:87]: yeah, that's a good idea, i'll change this to be just "train".

Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py]: Made this into just "train"

File: python/ray/ml/config.py [hunk 22]
```diff
@@ -1,13 +1,65 @@
...
+    def num_gpus_per_worker(self):
+        """The number of GPUs to set per worker."""
+        raise NotImplementedError
+
+    @property
+    def additional_resources_per_worker(self):
+        """Resources per worker, not including CPU or GPU resources."""
+        raise NotImplementedError
+
+    def get_placement_group_factory(self) -> "PlacementGroupFactory":
```
Comment (matthewdeng) [inline python/ray/ml/config.py:58]: nit: Would this be more intuitive if named `as_placement_group_factory` or `to_placement_group_factory`? Since this is essentially converting the object to another representation as opposed to retrieving an attribute.

File: python/ray/ml/train/data_parallel_trainer.py [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class DataParallelTrainer(Trainer):
+    """A Trainer for data parallel training.
+
+    You should subclass this Trainer if your Trainer follows SPMD (single program,
+    multiple data) programming paradigm- you want multiple processes to run the same
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:20]:
```suggestion
    multiple data) programming paradigm - you want multiple processes to run the same
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+    function, but on different data.
+
+    This Trainer runs the function ``train_loop_per_worker`` on multiple Ray
+    Actors. The provided datasets are automatically sharded to each Ray actor running
+    the training function.
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:25]: Datasets comment can be removed? Since we're only sharding the train dataset, and this is described below.
```suggestion
    Actors. 
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+            ...
+
+    If ``train_loop_per_worker`` accepts an argument, then
+    ``train_loop_config`` will be passed in as the argument. This is useful if you
+    want to tune the values in ``train_loop_config`` as hyperparameters.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    the "train" key), then it will be split into multiple dataset
+    shards that can be consumed by ``train_loop_per_worker``. All the other datasets
+    will not be sharded.
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:47]: Clarify what this means - every worker will retrieve the full dataset?

File: python/ray/ml/train/data_parallel_trainer.py [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+    :ref:`Ray Train function utils <train-api-func-utils>`.
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            # Report intermediate results for callbacks or logging.
+            train.report(...)
+
+            # Checkpoints the provided dict as restorable state.
+            train.save_checkpoint(...)
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:59]: nit: This will still be `kwargs` instead of a single input dict right?

File: python/ray/ml/train/data_parallel_trainer.py [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+            train.get_dataset_shard("my_dataset")
+
+            # Total number of workers executing training.
+            train.get_world_size()
+
+            # Rank of this worker.
+            train.get_world_rank()
+
+            # Rank of the worker on the current node.
+            train.get_local_rank()
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:74]: For consistency:
```suggestion
            # Returns the total number of workers executing training.
            train.get_world_size()

            # Returns the rank of this worker.
            train.get_world_rank()

            # Returns the rank of the worker on the current node.
            train.get_local_rank()
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 28]
```diff
@@ -0,0 +1,184 @@
...
+                from ray import train
+
+                def train_loop_for_worker():
+                    dataset_shard_for_this_worker = train.get_dataset_shard("train")
+
+                    assert len(dataset_shard_for_this_worker) == 1
+
+                train_dataset = ray.data.from_items([1, 2, 3])
+                assert len(train_dataset) == 3
+                trainer = DataParallelTrainer(scaling_config={"num_worker": 3},
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:91]:
```suggestion
                trainer = DataParallelTrainer(scaling_config={"num_workers": 3},
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 29]
```diff
@@ -0,0 +1,184 @@
...
+        Example:
+            .. code-block:: python
+
+                import ray
+                from ray import train
+
+                def train_loop_for_worker():
+                    dataset_shard_for_this_worker = train.get_dataset_shard("train")
+
+                    assert len(dataset_shard_for_this_worker) == 1
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:87]: Can be done in a follow-up PR, but I don't think this example sufficiently answers the question. 😅

File: python/ray/ml/train/data_parallel_trainer.py [hunk 30]
```diff
@@ -0,0 +1,184 @@
...
+            def __init__(self, train_loop_per_worker, my_backend_config:
+                MyBackendConfig, **kwargs):
+
+                super().__init__(train_loop_per_worker, my_backend_config, **kwargs)
+
+    Args:
+        train_loop_per_worker: The training function to execute.
+            This can either take in no arguments or a ``config`` dict.
+        train_loop_config: Configurations to pass into
+            ``train_func`` if it accepts an argument.
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:157]:
```suggestion
            ``train_loop_per_worker`` if it accepts an argument.
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 31]
```diff
@@ -0,0 +1,184 @@
...
+        backend_config: Used to specify which backend to setup on the
+            workers enable distributed communication, for example torch or
+            horovod. If no Backend should be set up, then set this to None.
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:160]:
```suggestion
        backend_config: Configuration for setting up a Backend (e.g. Torch,
            Tensorflow, Horovod) on each worker to enable distributed
            communication. If no Backend should be set up, then set this to None.
```

File: python/ray/ml/train/data_parallel_trainer.py [hunk 32]
```diff
@@ -0,0 +1,184 @@
...
+        scaling_config: Configuration for how to scale data parallel training.
+        run_config: Configuration for the execution of the training run.
+        datasets: Any Ray Datasets to use for training. Use
+            the key "train" to denote which dataset is the training
+            dataset. If a ``preprocessor`` is provided and has not already been fit,
+            it will be fit on the training dataset. All datasets will be transformed
+            by the ``preprocessor`` if one is provided.
+        preprocessor: A ray.ml.preprocessor.Preprocessor to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A checkpoint to resume training from.
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:170]: Any concerns if we just rename this to `checkpoint` for simplicity?

File: python/ray/ml/train/data_parallel_trainer.py [hunk 33]
```diff
@@ -0,0 +1,184 @@
...
+from ray.ml.checkpoint import Checkpoint
+from ray.train import BackendConfig
+from ray.util.annotations import DeveloperAPI
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class DataParallelTrainer(Trainer):
+    """A Trainer for data parallel training.
```
Comment (matthewdeng) [inline python/ray/ml/train/data_parallel_trainer.py:17]: Just wondering, are these docs being compiled anywhere? Or will that be verified in a separate PR?

**Review by matthewdeng — APPROVED** [review]
Mostly just minor comments, API looks good to me!

File: python/ray/ml/train/data_parallel_trainer.py (hunk 33 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py:17]: That will be in a separate PR.

File: python/ray/ml/train/data_parallel_trainer.py (hunk 32 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py:170]: I believe this was previously discussed and the decision was to go with `resume_from_checkpoint`? In any case, I am going to keep as is for now. Please bring this up with the broader group for an API change.

File: python/ray/ml/train/data_parallel_trainer.py (hunk 29 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py:87]: Can you clarify more here?

File: python/ray/ml/train/data_parallel_trainer.py (hunk 25 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py]: Updated!

File: python/ray/ml/train/data_parallel_trainer.py (hunk 26 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py]: yeah...I think for now. Changed.

File: python/ray/ml/config.py (hunk 22 above)
Comment (amogkam) [reply python/ray/ml/config.py]: Changed

File: python/ray/ml/trainer.py (hunk 21 above)
Comment (ericl) [reply python/ray/ml/trainer.py:128]: Got it

File: python/ray/ml/train/data_parallel_trainer.py (hunk 20 above)
Comment (amogkam) [reply python/ray/ml/train/data_parallel_trainer.py:41]: eh I think I'll leave this for now...Ray Train anyways has support for both.

Comment (ericl) [reply python/ray/ml/train/data_parallel_trainer.py:41]: Ok

**Review by krfricke — APPROVED** [review]
LGTM
